### PR TITLE
Add multi-access

### PR DESCRIPTION
**Pull request by SparklingComet**
*Will add a description later*

### CODE REVIEW

File: src/main/java/org/shanerx/tradeshop/Utils.java
Comment (KillerOfPie) [inline src/main/java/org/shanerx/tradeshop/Utils.java:24]: If we are importing all of bukkit shouldn't we get rid of the rest of the imports?

File: src/main/java/org/shanerx/tradeshop/admin/AdminEventListener.java
Comment (KillerOfPie) [inline src/main/java/org/shanerx/tradeshop/admin/AdminEventListener.java]: Why did we lose this return?

File: src/main/java/org/shanerx/tradeshop/admin/AdminEventListener.java
Comment (KillerOfPie) [inline src/main/java/org/shanerx/tradeshop/admin/AdminEventListener.java]: I want to play with this a bit to make it also show the name of the inventory as "Shop" or "Chest" when they open it

File: src/main/java/org/shanerx/tradeshop/admin/AdminEventListener.java
Comment (KillerOfPie) [inline src/main/java/org/shanerx/tradeshop/admin/AdminEventListener.java]: We should either have a method that  does both, or have members return all members and owners?

File: src/main/java/org/shanerx/tradeshop/bitrade/BiShopCreateEventListener.java
Comment (KillerOfPie) [inline src/main/java/org/shanerx/tradeshop/bitrade/BiShopCreateEventListener.java]: This can  go

File: src/main/java/org/shanerx/tradeshop/bitrade/BiShopCreateEventListener.java
Comment (KillerOfPie) [inline src/main/java/org/shanerx/tradeshop/bitrade/BiShopCreateEventListener.java]: do we want a method for checking these as well, or should we make the isTradeShop methods ignore case so they find these too?

File: src/main/java/org/shanerx/tradeshop/bitrade/BiShopCreateEventListener.java
Comment (KillerOfPie) [inline src/main/java/org/shanerx/tradeshop/bitrade/BiShopCreateEventListener.java]: should be getCreateBiPerm

File: src/main/java/org/shanerx/tradeshop/bitrade/BiShopCreateEventListener.java
Comment (KillerOfPie) [inline src/main/java/org/shanerx/tradeshop/bitrade/BiShopCreateEventListener.java]: move this into the try and have it throw an exception so we dont have to do the setlines and no chest message twice

**Review by KillerOfPie — CHANGES_REQUESTED** [review]

File: src/main/java/org/shanerx/tradeshop/admin/AdminEventListener.java
Comment (SparklingComet) [reply src/main/java/org/shanerx/tradeshop/admin/AdminEventListener.java]: The method's return type is void and it was the last statement of the if block. After that there is only an else block, so no need for it.

File: src/main/java/org/shanerx/tradeshop/admin/AdminEventListener.java
Comment (SparklingComet) [reply src/main/java/org/shanerx/tradeshop/admin/AdminEventListener.java]: What exactly do you mean by that?

File: src/main/java/org/shanerx/tradeshop/admin/AdminEventListener.java
Comment (SparklingComet) [reply src/main/java/org/shanerx/tradeshop/admin/AdminEventListener.java]: Will add a method that returns all members/owners

File: src/main/java/org/shanerx/tradeshop/bitrade/BiShopCreateEventListener.java
Comment (SparklingComet) [inline src/main/java/org/shanerx/tradeshop/bitrade/BiShopCreateEventListener.java]: Good catch, will remove.

File: src/main/java/org/shanerx/tradeshop/bitrade/BiShopCreateEventListener.java
Comment (SparklingComet) [reply src/main/java/org/shanerx/tradeshop/bitrade/BiShopCreateEventListener.java]: When I initially wrote the plugin (Idk how it looks like now) it automatically changed the case to [Trade]/[iTrade], so #equalsIgnoreCase was superficial. I am not sure how it works now.

File: src/main/java/org/shanerx/tradeshop/bitrade/BiShopCreateEventListener.java
Comment (SparklingComet) [reply src/main/java/org/shanerx/tradeshop/bitrade/BiShopCreateEventListener.java]: Uhm... I don't recall changing it... Good catch anyways, will fix.

File: src/main/java/org/shanerx/tradeshop/Utils.java
Comment (SparklingComet) [reply src/main/java/org/shanerx/tradeshop/Utils.java:24]: That was done by my IDE, didn't notice it. Wildcard imports are a *really* bad practice so I will fix this immediately.

File: src/main/java/org/shanerx/tradeshop/bitrade/BiShopCreateEventListener.java
Comment (SparklingComet) [reply src/main/java/org/shanerx/tradeshop/bitrade/BiShopCreateEventListener.java]: Will do.

File: src/main/java/org/shanerx/tradeshop/bitrade/BiShopCreateEventListener.java
Comment (SparklingComet) [reply src/main/java/org/shanerx/tradeshop/bitrade/BiShopCreateEventListener.java]: I thought so.

File: src/main/java/org/shanerx/tradeshop/bitrade/BiShopCreateEventListener.java
Comment (SparklingComet) [reply src/main/java/org/shanerx/tradeshop/bitrade/BiShopCreateEventListener.java]: maybe we should create a method in utils to set the lines... I'll add it to the TODO.

File: src/main/java/org/shanerx/tradeshop/bitrade/BiShopCreateEventListener.java
Comment (KillerOfPie) [reply src/main/java/org/shanerx/tradeshop/bitrade/BiShopCreateEventListener.java]: thats for the creation signs(so if a player doesnt cap them) its like that in all the creation files, but only for the first line

File: src/main/java/org/shanerx/tradeshop/bitrade/BiShopCreateEventListener.java
Comment (KillerOfPie) [reply src/main/java/org/shanerx/tradeshop/bitrade/BiShopCreateEventListener.java]: we can, would have to pass it the sign the trade type tho

File: src/main/java/org/shanerx/tradeshop/admin/AdminEventListener.java
Comment (KillerOfPie) [reply src/main/java/org/shanerx/tradeshop/admin/AdminEventListener.java]: k

File: src/main/java/org/shanerx/tradeshop/Utils.java
Comment (KillerOfPie) [inline src/main/java/org/shanerx/tradeshop/Utils.java]: Needs to use getState()

File: src/main/java/org/shanerx/tradeshop/Utils.java
Comment (KillerOfPie) [inline src/main/java/org/shanerx/tradeshop/Utils.java]: Needs to use getState()

**Review by KillerOfPie — CHANGES_REQUESTED** [review]
These need to get fixed before it works, also can you add the commands for adding owners and members. the break command is a good base to start with.

File: src/main/java/org/shanerx/tradeshop/Utils.java
Comment (SparklingComet) [reply src/main/java/org/shanerx/tradeshop/Utils.java]: Fixed in [this commit](https://github.com/SparklingComet/TradeShop/pull/19/commits/b4fd175bbadc031d7420a5ae2eeb638209826fc9)

File: src/main/java/org/shanerx/tradeshop/Utils.java
Comment (SparklingComet) [reply src/main/java/org/shanerx/tradeshop/Utils.java]: Fixed in [this commit](https://github.com/SparklingComet/TradeShop/pull/19/commits/b4fd175bbadc031d7420a5ae2eeb638209826fc9)

**Comment by SparklingComet** [comment]
I have added the addOwner and removeOwner methods in the last commit, but if someone tries to add someone who is already a member the target will be entered twice. There won't be any undesired effects, but it will take storage space, which is limited.
This will have to be fixed.

**Review by KillerOfPie — APPROVED** [review]

**Review by KillerOfPie — APPROVED** [review]
Looks good, gonna test everything real quick and see how many names we can put in

**Review by KillerOfPie — DISMISSED** [review]
Looks good i'm tired AF tho, have we added a check for total users on the chest yet? 
and also a check for if the user is already on the chest when we add them?
also players can be added that don't actually exist, are we going to check for that or just leave it to the players, there are no errors if a name is incorrect.

**Comment by SparklingComet** [comment]
Writing from my phone, so I cannot make any changes right now.

> Looks good i'm tired AF tho, have we added a check for total users on the chest yet?
and also a check for if the user is already on the chest when we add them?
also players can be added that don't actually exist, are we going to check for that or just leave it to the players, there are no errors if a name is incorrect.

getShopMembers() returns all members, regardless of whether they're owners or not. You can distinguish owners from members by checking whether getShopOwners also contains that OfflinePlayer.

There is no check like that. I believe it was added to the todo list, but it may have been accidentally moved to 'Completed' by me. We should definitely add it before merging, if you can do that (I cannot), that would be excellent.

Lastly, that can stay as it is, at least in my opinion. I also planned a /ts who commad that shows owners an members, so that players can double check who they added anyway.


Finally, I think there are other todo entries under 'Planned', but I may be wrong. Since I won't be home for a while,
I can only review changes but cannot actually commit, so, if you feel like adding it, I hardly have any complaints. Don't feel obligated to, though.

**Comment by KillerOfPie** [comment]
getShopMembers only returns members now, if you want everything I added a getShopUsers method,
and ill start working on the stuff but if you can get 1.5.2 out it would be nice

File: src/main/java/org/shanerx/tradeshop/Utils.java
Comment (SparklingComet) [inline src/main/java/org/shanerx/tradeshop/Utils.java:444]: I am sure this used to be different. The if check should be removed before merging. Another method returning both owners and members is useless.

**Comment by SparklingComet** [comment]
> getShopMembers only returns members now, if you want everything I added a getShopUsers method,
and ill start working on the stuff but if you can get 1.5.2 out it would be nice

I agree about the 1.5 part.

Edit: closed accidentally

**Comment by KillerOfPie** [comment]
had to remake the 1.6 branch to catch it up to master

**Comment by KillerOfPie** [comment]
This is working at this point there is a problem that must be addressed before it can be released though.
If a player that is not a owner on the shop puts a sign next to the chest it will add them to the chest(or make it theirs)

**Comment by SparklingComet** [comment]
Sounds nasty, but should be easy and quick to fix. Just check whether getShopSign returns a valid sign and, if yes, cancel the event with a kind message.
Is it safe to assume that the new patches are working?

**Comment by SparklingComet** [comment]
> Looks good, gonna test everything real quick and see how many names we can put in

Anything new about that? We need to implement a limit for that.

**Comment by KillerOfPie** [comment]
@SparklingComet already in max default users is 5

**Comment by KillerOfPie** [comment]
@SparklingComet problem needs to get fixed for double chests too

**Comment by SparklingComet** [comment]
I'll add a short checklist to track stuff to change/fic:
- [x] Stop non-owners to place (trade)signs on shop chests
- [x] Locking of doublechests needs to be fixed (?) (Found it in the todo list)

Once that's done and tested, I will merge this.

**Comment by SparklingComet** [comment]
I'm merging this into ``1.6`` because this is going beyond this PR's scope.

**Comment by SparklingComet** [comment]
All fixes need to be pushed to ``1.6-fixes``. (Branch is a copy of ``1.6``). Also, changes introduced in ``1.5.2``  need to be implemented.

**Comment by KillerOfPie** [comment]
@SparklingComet changes introduced in 1.5.2 were already merged and all errors from the were fixed

**Comment by SparklingComet** [comment]
@KillerOfPie excellent, I assume it had been tested and everything is working.

I am referencing #21 as it contains fixes for this PR.

**Comment by KillerOfPie** [comment]
Si

**Pull request merged**
